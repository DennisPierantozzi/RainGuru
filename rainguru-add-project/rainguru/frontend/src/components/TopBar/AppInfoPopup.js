import React from 'react';

/**
 * Generates a div for a popup where the content will be rendered in later.
 * @param props contains state of a particular popup, method to change the state, and text to display
 * @returns a div for a sidebar popup.
 */
function AppInfoPopup(props) {

    return (props.state) ? (
        <div className="popup" data-testid="AppInfoPopupContainer" onClick={(e) => props.changePopup(e, false)}>
            <div className="popup-box" data-testid="AppInfoPopupBox" onClick={(e) => props.changePopup(e, true)}>
                <div className="popup-content">
                    <div className="popup-information">
                        {props.children}
                    </div>
                    <div className="popup-close">
                        <button className="close-button" data-testid="AppInfoCloseButton" onClick={(e) => props.changePopup(e, false)}>
                            <img className="ratioImage" src="../../../static/images/3x1.png"/>
                            <p>Close</p>
                        </button>
                    </div>
                </div>
            </div>
        </div>
    ) : "";
}

export default AppInfoPopup