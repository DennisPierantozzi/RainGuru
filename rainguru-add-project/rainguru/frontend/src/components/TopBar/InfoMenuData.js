import React from 'react';
import { SiGnuprivacyguard } from 'react-icons/si';
import { CgDanger } from 'react-icons/cg';
import { HiInformationCircle } from 'react-icons/hi';
import { FaBook } from 'react-icons/fa';
import {IoMdPeople} from 'react-icons/io';

/**
 * Contains data to be displayed in a sidebar menu and sidebar popup.
 *  title: name of the sidebar menu
 *  icon: icon to be displayed in the sidebar menu next to title
 *  className: className that will be used to style them
 *  id: id for identifying each element in this SideBarData
 *  popupText: text to be displayed when the menu is clicked
 */
export const InfoMenuData = [
    {
        title: "Application information",
        icon: <HiInformationCircle className="sidebar-element-icon" size="1em"/>,
        className: "SideBar-item Bottom-menu-item",
        id:"application-info",
        popupText:
        <div class="popup-text" data-testid="application-info-popuptext-sidebardata">
            <div id="background" className='popup-text-box'>
            <h3><span>Background</span></h3>
            <p>The RainGuRu is a deep-learning rainfall forecasting model developed by TU Delft researchers and HKV
                (<a href="https://www.hkv.nl/">https://www.hkv.nl/</a>)  with the help of seed funding from the TU Delft
                Safety & Security Institute.
                The model uses a gated recurrent convolutional neural network known as Traj-GRU, first proposed in 2017
                by Shi et al.  (<a href="https://arxiv.org/abs/1706.03458">https://arxiv.org/abs/1706.03458</a>) and further developed and adapted for the Netherlands
              by TU Delft master students Eva van der Kooij (2021) and Diewertje Dekker (2022). The version shown on
                this website uses real-time KNMI radar data and a custom loss function designed to improve performance
                in times of heavy rain.
            </p>
            </div>
            <div id="app-info" className='popup-text-box'>
                <h3><span>Application information</span></h3>
                <p>This application was created from scratch in less than 8 weeks by 5 TU Delft Bachelor students in
                    computer science as part of the course “CSE2000 Software Project – Q4 (2021-2022)”. The application was
                    developed using Django <a href="https://www.djangoproject.com">(https://www.djangoproject.com)</a> and
                    React <a href="https://reactjs.org">(https://reactjs.org)</a> framework
                    for back-end and front-end respectively.
                    Furthermore, Leaflet <a href="https://leafletjs.com/">(https://leafletjs.com)</a> library has been used to
                    achieve interactivity of the map and finally the application
                    is currently running on the TU Delft virtual servers. </p><br/>
                <p>The new version of RainGuru, developed by Dennis Pierantozzi during his Erasmus+ Traineeship experience at TU Delft University, 
                    features a complete redesign and new functionality to display data through the use of a graph and map. 
                    The overall performance of the application has also been improved through changes made to the database and backend.
                </p>
                <h4><span>Interested in learning more?</span></h4>
                <p>RainGuRu is still in development and we hope to be able to provide newer, better versions soon.
                    If you are interested in contributing or want to learn more, please contact the principal investigator: <br/>
                    Dr. M.A. (Marc) Schleiss, Dept. of Geoscience & Remote Sensing, m.a.schleiss [at] <a href="https://www.tudelft.nl/">tudelft.nl</a>.</p>
            </div>
        </div>
    },
    {
        title: "Contributors",
        icon: <IoMdPeople className="sidebar-element-icon" size="1em"/>,
        className: "SideBar-item Bottom-menu-item" ,
        id: "contributor",
        popupText:
            <div class="popup-text" data-testid="contributor-popuptext-sidebardata">
                <div id="scientific-contributor" className='popup-text-box'>
                    <h3><span>Scientific contributors</span></h3>
                    <ul>
                        <li>Dr. Marc Schleiss, TU Delft</li>
                        <li>Dr. Francesco Fioranelli, TU Delft</li>
                        <li>Dr. Riccardo Taormina, TU Delft</li>
                        <li>Eva van der Kooij, TU Delft</li>
                        <li>Diewertje Dekker, TU Delft</li>
                        <li>Dr. Mattijn van Hoek, HKV</li>
                        <li>Ir. Dorien Lugt, HKV</li>
                    </ul>
                </div>
                <div id="development-team" className='popup-text-box'>
                    <div>
                        <h3><span>Initial software development team</span></h3>
                        <ul>
                            <li>Milan de Koning, TU Delft</li>
                            <li>Nikola Nachev, TU Delft</li>
                            <li>Thijs Penning, TU Delft</li>
                            <li>Kanta Tanahashi, TU Delft</li>
                            <li>Mike Raave, TU Delft</li>
                            <li>Dr.Elvin Isufi, TU Delft</li>
                            <li>Radu Gaghi, TU Delft</li>
                        </ul>
                    </div>
                    <div>
                        <h3><span>Version 2.0 full-stack developer</span></h3>
                        <ul>
                            <li>Dennis Pierantozzi, Erasmus+ Traineeship at TU Delft</li>
                        </ul>
                    </div>
                </div>
            </div>
    },
    {
        title: "Disclaimer & Liability",
        icon: <CgDanger className="sidebar-element-icon" size="1em"/>,
        className: "SideBar-item Bottom-menu-item",
        id: "disclaimer",
        popupText:
            <div class="popup-text" data-testid="disclaimer-popuptext-sidebardata">
                <div id="disclaimer" className='popup-text-box'>        
                        <h3><span>Disclaimer & Liability</span></h3>
                        <p>RainGuRu is a non-profit, curiosity driven research project. None of the authors, contributors and
                    institutions behind RainGuRu can be held accountable for the decisions taken by the users on the basis of
                    the presented forecasts. This includes any direct and / or indirect damage, such as loss of profit, loss of
                    business, loss of reputation, errors or omissions in the data, non-availability of data/app, defects,
                    viruses or malfunctioning of personal devices as a result of using this software.</p>
                </div>
            </div>
    },
    {
        title: "Acknowledgement",
        icon: <FaBook className="sidebar-element-icon" size="1em"/>,
        className: "SideBar-item Bottom-menu-item",
        id:"acknowledgement",
        popupText:
        <div class="popup-text" data-testid="acknowledgement-popuptext-sidebardata">
            <div id="acknowledgement" className='popup-text-box'>
                <h3><span>Acknowledgements</span></h3>
                <p>The team behind RainGuRu acknowledges the Dutch weather service KNMI for providing the real-time
                    radar 5-minute reflectivity composites needed to generate the forecasts. These data are provided
                    under the 'OpenData' policy of the Dutch government through the KNMI data platform
                    (<a href="https://dataplatform.knmi.nl/">https://dataplatform.knmi.nl/</a>)</p>
                <p>Special thanks to the ICT department of the Faculty of Civil Engineering and Geoscience at TU Delft
                    for hosting and managing the virtual server that hosts the RainGuRu application.</p>
                <p>Finally, we would like to thank the TU Delft Safety & Security Institute for providing the
                    seed funds that helped develop and implement RainGuRu.</p>
            </div>
            <div id="references" className='popup-text-box'>
                <h3><span>References</span></h3>
                <ul>
                    <li>Link to repository: <br/> <a href="https://gitlab.ewi.tudelft.nl/cse2000-software-project/2021-2022-q4/cluster-13/tu-delft-rainguru/tu-delft-rainguru.git">
                        https://gitlab.ewi.tudelft.nl/cse2000-software-project/2021-2022-q4/cluster-13/tu-delft-rainguru/tu-delft-rainguru.git</a></li>
                    <li>Van der Kooij, Eva (2021): Nowcasting heavy precipitation in the Netherlands: a deep learning approach:
                        <br/><a href="https://repository.tudelft.nl/islandora/object/uuid:536b1a77-625c-4476-9354-4d5b259a1080">
                        https://repository.tudelft.nl/islandora/object/uuid:536b1a77-625c-4476-9354-4d5b259a1080</a></li>
                    <li>Dekker, Diewertje (2022)</li>
                </ul>
            </div>
        </div>
    }
]