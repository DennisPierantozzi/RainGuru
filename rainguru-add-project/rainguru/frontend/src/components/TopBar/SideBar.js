import React from 'react';
import PastDataSelector from "./PastDataSelector";

export default function SideBar({pastDataSelectorUpdateProp, loadingData, displayData, compareData, displayComparedData, setShowCompare, setShowSidebar}) {

    return (
        <div className="Sidebar-shown">
            <div className='container-sidebar-menu'>
                <div className="Menu-title">
                    <div>RainGuru <br/> <span className="Menu-subtitle">Weather forecast app</span></div>
                    <img src="../../static/images/TUDelft_logo_rgb.webp" width="10vw" height="13vh" alt="TUDelft Logo" />
                </div>
                <PastDataSelector updateProp={pastDataSelectorUpdateProp} loadingData={loadingData} displayData={displayData} compareData={compareData} 
                setShowSidebar={setShowSidebar}/>

                <div className='Bottom-privacy-statement'>
                    <p>This website does not store privacy sensitive information, except for the location of the user during usage if permitted.</p>
                </div>
            </div> 
        </div>
    )
}