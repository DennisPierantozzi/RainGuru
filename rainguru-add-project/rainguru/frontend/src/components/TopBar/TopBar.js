import React, {useState} from 'react';
import {FaBars} from 'react-icons/fa';
import {HiOutlineDatabase} from 'react-icons/hi';
import {ImLocation} from 'react-icons/im';
import {SideBarData} from "./SideBarData";
import AppInfoPopup from "./AppInfoPopup";
import Map from "../Map/Map";
import PastDataSelector from "./PastDataSelector";

export default function TopBar({pastDataSelectorUpdateProp, loadingData, displayData}) {

    /**
     * initializing a state for sidebar and function to update it. Initially, state is set to false.
     * This state is to keep track of whether the sidebar is open or not
     */
    const [sidebar, setSidebar] = useState(false)

    /**
     * toggles the value to the opposite state.
     */
    const showSidebar = () => setSidebar(!sidebar)

    /**
     * displays precipitation information over user's current location and also leave a marker at that location.
     */
    const showCurrentLocation = () => {
        Map.map.locate();
    }

    /**
     * displays precipitation information at TU Delft and also marks the location.
     */
    const showTUDelftLocation = () => {
        const latlngPoint = new L.LatLng(52.0022,4.3736);
        Map.map.fireEvent('click',{latlng:latlngPoint});
    }

    /**
     * closes the pastData menu.
     */
    const closeMenus = () => {
        setPastDataSelector(false);
    }

    /**
     * initializing a state for pastDataSelector and function to update it. Initially, state is set to false.
     * This state is to keep track of whether pastData menu is open or not.
     */
    const [pastDataSelector, setPastDataSelector] = useState(false);

    /**
     * sets pastDataSelector state depending on the current state and closes menu if the current state is false.
     */
    const showPastDataSelector = () => {
        if (!pastDataSelector) {
            closeMenus();
            setPastDataSelector(true);
        } else {
            setPastDataSelector(false);
        }
    }

    /**
     * Initializing five states and functions to update them. Initial states are set to false for all of them.
     * These states are to keep track of whether or not a popup is displayed.
     */
    const [appInfoPopup, setAppInfoPopup] = useState(false)
    const [contributorPopup, setContributorPopup] = useState(false)
    const [disclaimerPopup, setDisclaimerPopup] = useState(false)
    const [privacyPopup, setPrivacyPopup] = useState(false)
    const [acknowledgementPopup, setAcknowledgementPopup] = useState(false);

    /**
     * Updates application information popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeAppInfoPopup = (e, bool) => {
        e.stopPropagation();
        setAppInfoPopup(bool);
    }

    /**
     * Updates disclaimer and liability popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeDisclaimerPopup = (e, bool) => {
        e.stopPropagation();
        setDisclaimerPopup(bool);
    }

    /**
     * Updates privacy statement popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changePrivacyPopup = (e, bool) => {
        e.stopPropagation();
        setPrivacyPopup(bool);
    }

    /**
     * Updates contributor popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeContributorPopup = (e, bool) => {
        e.stopPropagation();
        setContributorPopup(bool);
    }

    /**
     * Updates acknowledgement popup state.
     * @param e event parameter
     * @param bool state to be updated to
     */
    const changeAcknowledgementPopup = (e, bool) => {
        e.stopPropagation();
        setAcknowledgementPopup(bool);
    }

    /**
     * Renders the content of the TopBar div at the top of the screen. This also contains sidebar.
     * @returns the rendered content of the topbar div.
     */
    return (
        <div className="topBarContents" data-testid="topBar-container">
            <div className="Sidebar" data-testid="sidebar-menu">
                <img className="ratioImage" src="../../../static/images/1x1.png"/>
                <FaBars className="Sidebar-icon" onClick={showSidebar} size="60%"/>
            </div>
            <div className="Location">
                <img className="ratioImage" src="../../../static/images/1x1.png"/>
                <ImLocation className="Location-icon" data-testid="user-location" onClick={showCurrentLocation} size="60%"/>
            </div>
            <div className="TUDelft-logo">
                <img className="ratioImage" src="../../../static/images/1x1.png"/>
                <img src="../../../static/images/TUDelftButton.png" alt="TU Delft logo"
                     className="TUDelft-logo-icon" data-testid="tudelft-location" onClick={showTUDelftLocation}></img>
            </div>
            <div className="Searchbar" id="locationField"></div>

            <nav className={sidebar ? "Sidebar-shown" : "hideElement"}>
                <ul className="Sidebar-items">
                    <li className={pastDataSelector ? "active-menu-item" : "SideBar-item"} onClick={showPastDataSelector}>
                        <span className="item-icon">
                            <img className="ratioImage" src="../../../static/images/1x1.png"/>
                            <HiOutlineDatabase className="sidebar-element-icon" size="50%"/>
                        </span>
                        <span className="item-title">Display previous data</span>
                    </li>
                    <li className={pastDataSelector ? "open-menu" : "hideElement"}>
                        <PastDataSelector updateProp={pastDataSelectorUpdateProp} loadingData={loadingData} displayData={displayData}/>
                    </li>
                    <li className={pastDataSelector ? "hideElement" : "sidebar-splitter"} id="sidebar-splitter"></li>
                    {SideBarData.map((item, index) => {
                        return (
                            <li key={index} className={item.className} data-testid={item.id}
                                onClick={item.title === "Application information" ? (e) => changeAppInfoPopup(e, true) :
                                    item.title === "Contributors" ? (e) => changeContributorPopup(e, true) :
                                    item.title === "Disclaimer & Liability" ? (e) => changeDisclaimerPopup(e, true) :
                                    item.title === "Acknowledgement" ? (e) => changeAcknowledgementPopup(e, true) :
                                        (e) => changePrivacyPopup(e, true)}>
                                <span className="item-icon">
                                    <img className="ratioImage" src="../../../static/images/1x1.png"/>
                                    {item.icon}
                                </span>
                                <span className="item-title">{item.title}</span>
                                <div className="popup-container" data-testid={item.id + "-popup-container"}>
                                    <AppInfoPopup
                                        state={item.title === "Application information" ? appInfoPopup : item.title === "Disclaimer & Liability" ? disclaimerPopup:
                                            item.title === "Contributors" ? contributorPopup : item.title === "Acknowledgement" ? acknowledgementPopup : privacyPopup}
                                        changePopup={item.title === "Application information" ? changeAppInfoPopup :
                                            item.title === "Contributors" ? changeContributorPopup:
                                            item.title === "Disclaimer & Liability" ?  changeDisclaimerPopup:
                                            item.title === "Acknowledgement" ? changeAcknowledgementPopup: changePrivacyPopup}>
                                        <div id="popup-text"data-testid={item.id + "-popuptext-topbar"}>{item.popupText}</div>
                                    </AppInfoPopup>
                                </div>
                            </li>
                        )
                    })}
                </ul>
            </nav>
        </div>
    )
}
