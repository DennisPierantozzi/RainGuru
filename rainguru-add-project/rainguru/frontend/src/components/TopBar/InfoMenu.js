import React, {useState, useEffect } from 'react';
import {InfoMenuData} from "./InfoMenuData";

function InfoMenu () {

    const [textToShow, setTextToShow] = useState("");
    const [previousItem, setPreviousItem] = useState("");

    // set the first info section as the default one
    useEffect(() => {
        setTextToShow(InfoMenuData[0].popupText);
    }, []);


    /**
     * handle the click on the item to show different sections of the menu
     *
     * @param item the item clicked by the user
     */
    const handleClickInfo = (item) => {
        document.getElementById(item.title).style.color = "#0076C2";
        if(previousItem != "") {document.getElementById(previousItem).style.color = "#000";}
        setPreviousItem(item.title);
        setTextToShow(item.popupText);
    }

    return(
        <div id="infoMenu" className="Sidebar-shown">
            <div className='container-sidebar-menuInfo'>
                <div className="Menu-title">
                        <div>RainGuru <br/> <span className="Menu-subtitle">Weather forecast app</span></div>
                        <img src="../../static/images/TUDelft_logo_rgb.webp" width="10vw" height="13vh" alt="TUDelft Logo" />
                </div>
                <div className="contentsMenuInfo">
                    <div className="list-menuInfo">
                        {InfoMenuData.map((item,i) => <div id={item.title} className="elements-list-menuInfo" onClick={() => {handleClickInfo(item)}}>
                        <div class="item-list-display">{item.icon}</div> <div className='item-list-display'>{item.title}</div></div>)}
                    </div>
                    {textToShow}
                </div>
                <div className='Bottom-privacy-statement'>
                    <p>This website does not store privacy sensitive information, except for the location of the user during usage if permitted.</p>
                </div>
            </div>
        </div>
    );
}

export default InfoMenu;