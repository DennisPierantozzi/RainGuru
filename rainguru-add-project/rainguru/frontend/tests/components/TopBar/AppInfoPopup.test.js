import renderer from "react-test-renderer";
import AppInfoPopup from "../../../src/components/TopBar/AppInfoPopup";
import {fireEvent, render, screen} from "@testing-library/react";
import React from "react";
import TopBar from "../../../src/components/TopBar/TopBar";

jest.mock("../../../src/components/TopBar/PastDataSelector")

describe("AppInfoPopup tests", () => {

    it("matches snapshot", () => {
        const state = renderer.create(<AppInfoPopup state={true} changePopup={jest.fn((e, bool) => e.stopPropagation())} children={"hi"}/>).toJSON();
        expect(state).toMatchSnapshot();
    });


    it("renders AppInfoContainer", () => {
        const {getByTestId} = render(<AppInfoPopup state={true} changePopup={jest.fn((e, bool) => e.stopPropagation())} children={"hi"}/>);
        expect(getByTestId("AppInfoPopupContainer")).toBeTruthy();
    });

    it("clicking on AppInfoContainer test", () => {
        const mockChangePopup = jest.fn();
        const {getByTestId} = render(<AppInfoPopup state={true} changePopup={mockChangePopup} children={"hi"}/>);
        const AppInfoPopupContainer = getByTestId("AppInfoPopupContainer");
        fireEvent.click(AppInfoPopupContainer);
        expect(mockChangePopup).toHaveBeenCalledTimes(1);
    });

    it("clicking on AppInfoContainer test", () => {
        const mockChangePopup = jest.fn((e, bool)=> e.stopPropagation());
        const {getByTestId} = render(<AppInfoPopup state={true} changePopup={mockChangePopup} children={"hi"}/>);
        const AppInfoPopupBox = getByTestId("AppInfoPopupBox");
        fireEvent.click(AppInfoPopupBox);
        expect(mockChangePopup).toHaveBeenCalledTimes(1);
    });

    it("clicking on AppInfoContainer test", () => {
        const mockChangePopup = jest.fn((e, bool)=> e.stopPropagation());
        const {getByTestId} = render(<AppInfoPopup state={true} changePopup={mockChangePopup} children={"hi"}/>);
        const AppInfoPopupContainer = getByTestId("AppInfoCloseButton");
        fireEvent.click(AppInfoPopupContainer);
        expect(mockChangePopup).toHaveBeenCalledTimes(1);
    });
});