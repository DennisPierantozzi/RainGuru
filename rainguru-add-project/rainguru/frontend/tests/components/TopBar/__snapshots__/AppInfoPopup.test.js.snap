// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInfoPopup tests matches snapshot 1`] = `
<div
  className="popup"
  data-testid="AppInfoPopupContainer"
  onClick={[Function]}
>
  <div
    className="popup-box"
    data-testid="AppInfoPopupBox"
    onClick={[Function]}
  >
    <div
      className="popup-content"
    >
      <div
        className="popup-information"
      >
        hi
      </div>
      <div
        className="popup-close"
      >
        <button
          className="close-button"
          data-testid="AppInfoCloseButton"
          onClick={[Function]}
        >
          <img
            className="ratioImage"
            src="../../../static/images/3x1.png"
          />
          <p>
            Close
          </p>
        </button>
      </div>
    </div>
  </div>
</div>
`;
